packages/flutter/lib/src/cupertino/search_field.dart

```diff
@@ -0,0 +1,291 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/widgets.dart';
+
+import 'colors.dart';
+import 'icons.dart';
+import 'text_field.dart';
+
+/// A [CupertinoTextField] that mimics the look and behavior of UIKit's
+/// `UISearchTextField`.
+///
+/// This control defaults to showing the basic parts of a `UISearchTextField`,
+/// like the 'Search' placeholder, prefix-ed Search icon, and suffix-ed
+/// X-Mark icon.
+///
+/// To control the text that is displayed in the text field, use the
+/// [controller]. For example, to set the initial value of the text field, use
+/// a [controller] that already contains some text such as:
+///
+/// {@tool snippet}
+///
+/// ```dart
+/// class MyPrefilledSearch extends StatefulWidget {
+///   @override
+///   _MyPrefilledSearchState createState() => _MyPrefilledSearchState();
+/// }
+///
+/// class _MyPrefilledSearchState extends State<MyPrefilledSearch> {
+///   TextEditingController _textController;
+///
+///   @override
+///   void initState() {
+///     super.initState();
+///     _textController = TextEditingController(text: 'initial text');
+///   }
+///
+///   @override
+///   Widget build(BuildContext context) {
+///     return CupertinoSearchTextField(controller: _textController);
+///   }
+/// }
+/// ```
+/// {@end-tool}
+///
+/// It is recommended to pass a [ValueChanged<String>] to both [onChanged] and
+/// [onSubmitted] parameters in order to be notified once the value of the
+/// field changes or is submitted by the keyboard:
+///
+/// /// {@tool snippet}
+///
+/// ```dart
+///
+///   @override
+///   Widget build(BuildContext context) {
+///     return CupertinoSearchTextField(
+///       onChanged: (value) {
+///         print("The text has changed to: " + value);
+///       },
+///       onSubmitted: (value) {
+///         print("Submitted text: " + value);
+///       },
+///     );
+///   }
+/// }
+/// ```
+/// {@end-tool}
+class CupertinoSearchTextField extends StatefulWidget {
+  /// Creates a [CupertinoTextField] that mimicks the look and behavior of UIKit's
+  /// `UISearchTextField`.
+  ///
+  /// Similar to [CupertinoTextField], to provide a prefilled text entry, pass
+  /// in a [TextEditingController] with an initial value to the [controller]
+  /// parameter.
+  ///
+  /// The [onChanged] parameter takes a [ValueChanged<String>] which is invoked
+  /// upon a change in the text field's value.
```

Comment:
nit: `onChanged` is called on user text input (it won't fire if you change `TextEditingController.value` programmatically).